packages/adblocker-extended-selectors/src/eval.ts

```diff
@@ -71,104 +71,247 @@ export function matches(element: Element, selector: AST): boolean {
+      }
+      break;
+    case '>':
+      for (const e of leftElements) {
+        elements.push(...Array.from(e.children).filter((child) => matches(child, selector.right)));
       }
-    } else if (selector.combinator === '~') {
-      for (const element2 of elements2) {
-        let sibling: Element | null = element2;
+      break;
+    case '~':
+      for (const e of leftElements) {
+        let sibling: Element | null = e;
         while ((sibling = sibling.nextElementSibling) !== null) {
-          if (matches(sibling, selector.right) === true) {
+          if (matches(sibling, selector.right)) {
             elements.push(sibling);
           }
         }
       }
-    } else if (selector.combinator === '+') {
-      for (const element2 of elements2) {
-        const nextElementSibling = element2.nextElementSibling;
-        if (nextElementSibling !== null && matches(nextElementSibling, selector.right) === true) {
-          elements.push(nextElementSibling);
+      break;
+    case '+':
+      for (const e of leftElements) {
+        const next = e.nextElementSibling;
+        if (next !== null && matches(next, selector.right)) {
+          elements.push(next);
         }
       }
-    }
-  } else if (selector.type === 'pseudo-class') {
-    // if (selector.name === 'upward') {
-    //   let n = Number(selector.argument);
-    //   console.log('upward', selector, n);
-    //   if (Number.isNaN(n) === false) {
-    //     if (n >= 1 && n < 256) {
-    //       let ancestor: Element | null = element;
-    //       while (ancestor !== null && n > 0) {
-    //         ancestor = ancestor.parentElement;
-    //         n -= 1;
-    //       }
-
-    //       if (ancestor !== null && n === 0) {
-    //         elements.push(element);
-    //       }
-    //     }
-    //   } else if (selector.argument !== undefined) {
-    //     const parent = element.parentElement;
-    //     if (parent !== null) {
-    //       const ancestor = parent.closest(selector.argument);
-    //       if (ancestor !== null) {
-    //         elements.push(ancestor);
-    //       }
-    //     }
-    //   }
-    // } else {
-    for (const subElement of element.querySelectorAll('*')) {
-      if (matches(subElement, selector) === true) {
-        elements.push(subElement);
+      break;
+  }
+
+  return elements;
+}
+
+export function querySelectorAll(element: Element, selector: AST): Element[] {
+  if (
+    selector.type === 'id' ||
+    selector.type === 'class' ||
+    selector.type === 'type' ||
+    selector.type === 'attribute'
+  ) {
+    return Array.from(element.querySelectorAll(selector.content));
+  }
+
+  if (selector.type === 'list') {
+    return selector.list.flatMap((s) => querySelectorAll(element, s));
+  }
+
+  if (selector.type === 'compound') {
+    return handleCompoundSelector(element, selector.compound);
+  }
+
+  if (selector.type === 'complex') {
+    return handleComplexSelector(element, selector);
+  }
```

Comment:
Do we want to ensure that returned element are unique here? I think it's possible that the same element might be returned more than once in some cases.